llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8392,154 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+// As mentioned, this does preparation work like inserting the bounds checks, its up to the caller caring about
+// inserting the OpLoad or OpStore. This also sets a flag that signals the main translation loop that we want to
+// continue inserting subsequent instructions at the final block instead of just going back to the parent block.
+// Currently, OpLoads will yield zero and OpStores will just omit the store, if bounds checks failed. This is up to the
+// caller. This implementation handles nested structs using arrays and 1...n-dimensional arrays. This returns if we need
+// to insert a bounds checks and pointers to the newly created basic blocks (which can be optimized away by subsequent
+// passes, if not needed).
+//
+// An example: We have the following sequence in LLVM IR:
+//
+// entry:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+//
+// which can cause a out of bounds access.
+//
+// However, we are converting it to a sequence like this:
+//
+// entry:
+//
+// %shouldLoad1 = icmp ult i32 runtimeIndex, localArray.dim1.upperBound
+// %shouldLoad2 = icmp ult i32 runtimeIndex * 2, localArray.dim2.upperBound
+// %shouldLoad = and i1 %shouldLoad1, %shouldLoad2
+// br i1 %shouldLoad, label %loadBlock, label %finalBlock
+//
+// loadBlock:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+// br label %finalBlock
+//
+// finalBlock:
+// %finalResult = phi ... <4 x float> [ zeroinitializer, %entry], [ %load, loadBlock ]
+// ... use %finalResult subsequently
+//
+SPIRVToLLVM::ScratchBoundsCheckInfo SPIRVToLLVM::prepareScratchBoundsCheck(SPIRVValue *toBeWrapped) {
+  constexpr ScratchBoundsCheckInfo nullValue{};
+  if (!getPipelineOptions()->enableScratchAccessBoundsChecks) {
+    return nullValue;
+  }
+
+  if (!toBeWrapped || toBeWrapped->getOpCode() != OpAccessChain) {
+    return nullValue;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = toBeWrapped->getType()->getPointerStorageClass();
+
+  // We don't want bounds checks if we are outside private or function storage class.
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return nullValue;
+  }
+
+  // Fetch the access chain behind the pointer the memory operation uses.
+  const auto accessChainBase = static_cast<SPIRVAccessChainBase *const>(toBeWrapped);
+
+  llvm::SmallVector<SPIRVValue *> indicesVec;
+
+  // Find any integer index, either for accessing the array or a struct member.
+  for (auto *const operand : accessChainBase->getIndices()) {
+    assert(operand->getType()->isTypeInt());
+    indicesVec.push_back(operand);
+  }
+
+  if (indicesVec.empty()) {
+    return nullValue;
+  }
+
+  ArrayRef<SPIRVValue *> indices = indicesVec;
+  SPIRVValue *currentIndex = nullptr;
+
+  auto consumeFirstIndex = [&currentIndex, &indices]() {
+    assert(!indices.empty());
+    currentIndex = indices[0];
+    indices = indices.drop_front();
+  };
```

Comment:
Alternatively, could we make this return the first index or nullptr?
This would make the loop something like:
`for (SPIRVValue *index = consumeFirstIndex(); index; consumeFirstIndex()) {`
or just
`while (SPIRVValue *index = consumeFirstIndex()) {`

Reply:
How about `for (const SPIRVValue *operand : accessChainBase->getIndices()) {`?
I think `indicesVec` is not needed at all?

Reply:
I cleaned up the whole loop a little bit, so we got rid of the `consumeFirstIndex` and all the other stuff.